src/main/java/com/zufar/onlinestore/security/signin/attempts/LoginAttemptManager.java

```diff
@@ -0,0 +1,59 @@
+package com.zufar.onlinestore.security.signin.attempts;
+
+import com.zufar.onlinestore.security.signin.attempts.entity.LoginAttemptEntity;
+import com.zufar.onlinestore.security.exception.AccountLockedException;
+import com.zufar.onlinestore.security.signin.attempts.repository.LoginAttemptRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDateTime;
+import java.util.Optional;
+
+@Service
+@RequiredArgsConstructor
+public class LoginAttemptManager {
+
+    private static final int MAX_LOGIN_ATTEMPTS = 5;
+    private static final int LOCKOUT_DURATION_MINUTES = 30;
+    private static final int INITIAL_LOGIN_ATTEMPTS_COUNT = 1;
+
+    private final LoginAttemptRepository loginAttemptRepository;
+
+    public void handleFailedLogin(final String userEmail) {
+        Optional<LoginAttemptEntity> existingAttempt = loginAttemptRepository.findByUserEmail(userEmail);
+
+        LoginAttemptEntity loginAttempt;
+        if (existingAttempt.isEmpty()) {
```

Comment:
What happens If 5 parallel requests come at the small time interval AND existingAttempt does not exist ?
5 parallel requests create 5 new loginAttempt objects: loginAttempt = new LoginAttemptEntity()
Then below you try to save all 5 objects:  loginAttemptRepository.save(loginAttempt);
First request completes succesfully but the remaining 4 will not be saved in the database due to a unique email constaint.
And arise 5 exceptions that are not handled.
Only one attempt was saved but there were 5 of them and we lost 4 bad attempts